### ISSUE TITLE
Unexpectedly low distances returned by '_C.point_face_dist_forward'

### ISSUE THREAD

**Issue by ehedlin** [facebookresearch/pytorch3d#1045]
I am trying to find the distance/index from the closest point to a mesh. When I run this through `_C.point_face_dist_forward` I am getting values that I am finding consistent on most of them (within a fraction of a percent from what I would expect) but around 1/10th of the values are too low. For example, in my mesh the face closest to the point `[ 0.0384, -0.1066, -0.2238]` is apparently 
`[[-0.0040, -0.0933, -0.0330],
   [-0.0003, -0.0958, -0.0513],
   [-0.0045, -0.0934, -0.0363]]`
but the returned distance is 0.0002 (The value for the distance returned by the function `_C.point_face_dist_forward` is the squared distance and this is after I un-square it). I am getting the distance from that point to that face to actually be 0.1771. 

Here is the code I am using to get the closest index/distance:

```
import pytorch3d
from pytorch3d import _C
from pytorch3d.structures import Meshes, Pointclouds

mesh1 = Meshes(verts=smpl_vertices, faces=smpl_faces)
pcd = Pointclouds(points=points.unsqueeze(0))

# packed representation for pointclouds
points = pcd.points_packed()  # (P, 3)
points_first_idx = pcd.cloud_to_packed_first_idx()
max_points = pcd.num_points_per_cloud().max().item()

# packed representation for faces
verts_packed = mesh1.verts_packed()
faces_packed = mesh1.faces_packed()
tris = verts_packed[faces_packed]  # (T, 3, 3)
tris_first_idx = mesh1.mesh_to_faces_packed_first_idx()
max_tris = mesh1.num_faces_per_mesh().max().item()

# point to face distance: shape (P,)
dists, idxs = _C.point_face_dist_forward(
    points, points_first_idx, tris, tris_first_idx, max_points
)
```

**Comment by gkioxari** [comment]
I just ran the following code:

```python
device = torch.device("cuda:0")
tris = torch.tensor([[-0.0040, -0.0933, -0.0330], [-0.0003, -0.0958, -0.0513], [-0.0045, -0.0934, -0.0363]], dtype=torch.float32, device=device).view(1, 3, 3)
point = torch.tensor([[0.0384, -0.1066, -0.2238]], dtype=torch.float32, device=device).view(1, 3)

dists, idxs = _C.point_face_dist_forward(
    cloud, torch.zeros((1,), device=device, dtype=torch.int64), verts, torch.zeros((1,), device=device, dtype=torch.int64), 1
)
```
And got for `dists`
```python
tensor([0.0314], device='cuda:0')
```

and `sqrt(0.0314) = 0.1772004515`

I don't have your full data to reproduce your code.

**Comment by ehedlin** [comment]
Interestingly it seems to do well with individual queries but with meshes with areas of higher density, it seems to get stuck for some reason. I am mainly interested in the returned index (of the nearest face) so that I can get the nearest point on the triangle and as a result, the point on the mesh nearest the query point. I have attached a visualization of the points for this one example here. The black is the query point, the blue is the nearest point that I have found (through my method which is painfully slow) and the red is the nearest point on the face whose index is returned (larger because it tends to hide in clumps of vertices). I find most of the time when my method disagrees with _C.point_face_dist_forward, it's when the latter returns a point somewhere on the ear. 
![Screen Shot 2022-02-07 at 8 37 19 PM](https://user-images.githubusercontent.com/23744775/152919893-45631789-8ec5-4dbe-a9ae-96f1c6f75608.png)
![test_9](https://user-images.githubusercontent.com/23744775/152919907-5df87d03-8880-4e5c-81ba-aff90123c408.png)
![test_8](https://user-images.githubusercontent.com/23744775/152919910-3e109dfa-cb69-430a-88c4-fc438d3a4987.png)
![test_7](https://user-images.githubusercontent.com/23744775/152919914-54ff022a-9033-41bb-a396-eb02a856adb8.png)
![test_6](https://user-images.githubusercontent.com/23744775/152919919-86eaaf56-a42a-43b9-92aa-0edf4f5f6474.png)
![test_5](https://user-images.githubusercontent.com/23744775/152919922-58676f7b-1034-4130-b8ad-0d002cc036b2.png)
![test_4](https://user-images.githubusercontent.com/23744775/152919925-f50b9eb7-2a51-4ce5-8df1-7fb3add1b63d.png)
![test_3](https://user-images.githubusercontent.com/23744775/152919929-de0c0b86-a387-4b25-a82f-39efc6a93d59.png)
![test_2](https://user-images.githubusercontent.com/23744775/152919932-31fa3344-cb18-4ea3-a382-6f08e9888de8.png)
![test_1](https://user-images.githubusercontent.com/23744775/152919937-232dc7b8-45f6-4bd8-9e0d-ab691081fa08.png)
![test_0](https://user-images.githubusercontent.com/23744775/152919941-e4f561ce-fa4b-4370-a168-17a771b0ca7e.png)

**Comment by ehedlin** [comment]
And here is my code given the input query points, mesh vertices, faces and indexes:
"""
def get_nearest_point_for_indeces(points, verts, faces, idxs):
    these_triangles = faces[idxs[0]]

    a = verts[these_triangles[:, 0]]
    b = verts[these_triangles[:, 1]]
    c = verts[these_triangles[:, 2]]
    

    # find the normal of the 3 points
    n = torch.cross(b - a, c - a)
    # normalize n
    n = n / torch.norm(n, dim=-1, keepdim=True)

    # find the centre points of the triangles
    p0 = (a+b+c)/3
    
    # find the directions between points and p0
    v = points-p0

    # the dist is the absolute value of the dot product between n and v
    dist = torch.abs(v.unsqueeze(-2)@n.unsqueeze(-1))[..., 0]

    # find the point on the triangle in the direction of the normal
    nearest_point = points - n*dist

    # make sure nearest_point stays within the triangle of determined by nearest_points
    # first check whether the point is to the left or right of each edge
    # if the point is to the left, project it to the line, otherwise leave it

    # inside = torch.ones(nearest_point.shape).cuda()
    

    for i, j in [[a, b], [b, c], [c, a]]:

        # find the nearest point from p0 to line
        point, _ = find_nearest_point_on_line_to_point(i, j, p0)

        outward_vector = point-p0

        this_vector = nearest_point-point
        # normalize this_vector
        this_vector = this_vector / torch.norm(this_vector, dim=-1, keepdim=True)
        outward_vector = outward_vector / torch.norm(outward_vector, dim=-1, keepdim=True)

        # if the dot product between outward_vector and this_vector is negative, project it to the line
        is_inside_triangle = (outward_vector.unsqueeze(-2)@this_vector.unsqueeze(-1))[..., 0] < 0

        # inside = torch.where(is_inside_triangle, inside, torch.zeros(inside.shape).cuda())

        projected_points, _ = find_nearest_point_on_line_to_point(i, j, nearest_point)
        nearest_point = torch.where(is_inside_triangle, nearest_point, projected_points)

    # find the corresponding point on the canonical mesh

    distance = torch.norm(nearest_point - points, dim=-1, keepdim=True)

    return nearest_point.unsqueeze(0), distance.unsqueeze(0)
"""

**Comment by gkioxari** [comment]
The only way for me to debug this is if you give me code and data. Without data there is nothing I can do. You are saying it has to do something with the density of the mesh, but the implementation is not related to any notion of density. 

If you can't share the data for me to debug this, you could share only part of the mesh in a drive folder and delete it right afterwards (e.g. the head only) so that I can reproduce this. I am happy to help you but without data I can't do much, I don't have magic powers :(

**Comment by ehedlin** [comment]
It gets stuck on the SMPL mesh. Here is the template mesh 

[smpl_uv.obj.zip](https://github.com/facebookresearch/pytorch3d/files/8085471/smpl_uv.obj.zip)

**Comment by gkioxari** [comment]
Here is the output I get and the code that I used to get there. Blue is the query point, red is the center point of the returned closest face with PyTorch3D. I don't see any issues. (For reference, this is the code we expect users to provide if they need help from us)

![smpl_example00](https://user-images.githubusercontent.com/4369065/154554693-5f5c6e78-09ef-425c-a260-a00b97453258.png)


```python
mesh = IO().load_mesh("/tmp/smpl_uv.obj", device=device)

# colors the human green!
color = torch.tensor([0.0, 1.0, 0.0], device=device).view(1, 1, 3).expand(-1, mesh._V, -1)
tex = TexturesVertex(color)
mesh = Meshes(verts=mesh.verts_list(), faces=mesh.faces_list(), textures=tex)

# get triangles
verts = mesh.verts_list()[0]
faces = mesh.faces_list()[0]
tris = verts[faces]

# query point
point = torch.tensor([[0.0684, 0.5266, 0.0638]], dtype=torch.float32, device=device).view(1, 3)

# point to tri distance
dists, idxs = _C.point_face_dist_forward(
    point, torch.zeros((1,), device=device, dtype=torch.int64), tris, torch.zeros((1,), device=device, dtype=torch.int64), 1
)

# center point of closest face
pp = tris[idxs[0]].mean(0)

# visualization
ico3 = ico_sphere(1, device=device)
vv, ff = ico3.get_mesh_verts_faces(0)
r = 0.01

vv1 = r * vv + pp.view(1, 3)
color1 = torch.tensor([1.0, 0.0, 0.0], device=device).view(1, 1, 3).expand(-1, vv1.shape[0], -1)
vv2 = r * vv + point.view(1, 3)
color2 = torch.tensor([0.0, 0.0, 1.0], device=device).view(1, 1, 3).expand(-1, vv2.shape[0], -1)

mesh1 = Meshes(verts=[vv1], faces=[ff], textures=TexturesVertex(color1))
mesh2 = Meshes(verts=[vv2], faces=[ff], textures=TexturesVertex(color2))

mesh_scene = join_meshes_as_scene([mesh, mesh1, mesh2])

IO().save_mesh(mesh_scene, "/tmp/smpl_uv_out.ply", colors_as_uint8=True)
```

**Comment by ehedlin** [comment]
Using your code, the attached mesh, and the point (0.1091, -1.1166, -0.0547) I ended up with results that look like this.

[debug.obj.zip](https://github.com/facebookresearch/pytorch3d/files/8093228/debug.obj.zip)
![Screen Shot 2022-02-17 at 4 51 40 PM](https://user-images.githubusercontent.com/23744775/154596539-da46538f-04e6-4cee-8e66-f317a89ed6fd.png)

**Comment by gkioxari** [comment]
I was able to reproduce this! The issue is with the fact that you have a very small face in your SMLP mesh. Small faces are handled with an `epsilon` check in the code, but yours is smaller than that and so it was not caught. The diff should land soon, but if you are impatient change the following from `1e-5` to `5e-3`

https://github.com/facebookresearch/pytorch3d/blob/59972b121d8c7bfc0e156b5ad5fcd77c11874178/pytorch3d/csrc/utils/geometry_utils.h#L654

https://github.com/facebookresearch/pytorch3d/blob/59972b121d8c7bfc0e156b5ad5fcd77c11874178/pytorch3d/csrc/utils/geometry_utils.cuh#L553

**Comment by ehedlin** [comment]
Ah, that clears it up. Thank-you very much :)

**Comment by gkioxari** [comment]
This https://github.com/facebookresearch/pytorch3d/commit/ee71c7c44786e41049829a5e461ce860d80a80c7 fixes it
